packages/hash/src/index.ts

```diff
@@ -2,62 +2,69 @@
 // Inspired by https://github.com/garycourt/murmurhash-js
 // Ported from https://github.com/aappleby/smhasher/blob/61a0530f28277f2e850bfc39600ce61d02b518de/src/MurmurHash2.cpp#L37-L86
 
+const encoder = new TextEncoder()
+
+// bufferLength must be a multiple of 4 to satisfy Int32Array constraints
+let bufferLength = 2 * 1024
+let buffer = new ArrayBuffer(bufferLength)
+let uint8View = new Uint8Array(buffer)
+let int32View = new Int32Array(buffer)
+
 export default function murmur2(str: string): string {
+  if (str.length > bufferLength) {
+    // buffer.resize() is only available in recent browsers, so we re-allocate
+    // a new buffer and views
+    bufferLength = str.length + (4 - str.length % 4)
+    buffer = new ArrayBuffer(bufferLength)
+
+    uint8View = new Uint8Array(buffer)
+    int32View = new Int32Array(buffer)
+  }
+
+  const length = encoder.encodeInto(str, uint8View).written;
+
   // 'm' and 'r' are mixing constants generated offline.
   // They're not really 'magic', they just happen to work well.
 
-  // const m = 0x5bd1e995;
-  // const r = 24;
+  const m = 0x5bd1e995;
+  const r = 24;
 
   // Initialize the hash
-
-  var h = 0
+  let h = 0
 
   // Mix 4 bytes at a time into the hash
 
-  var k,
-    i = 0,
-    len = str.length
-  for (; len >= 4; ++i, len -= 4) {
-    k =
-      (str.charCodeAt(i) & 0xff) |
-      ((str.charCodeAt(++i) & 0xff) << 8) |
-      ((str.charCodeAt(++i) & 0xff) << 16) |
-      ((str.charCodeAt(++i) & 0xff) << 24)
-
-    k =
-      /* Math.imul(k, m): */
-      (k & 0xffff) * 0x5bd1e995 + (((k >>> 16) * 0xe995) << 16)
-    k ^= /* k >>> r: */ k >>> 24
-
-    h =
-      /* Math.imul(k, m): */
-      ((k & 0xffff) * 0x5bd1e995 + (((k >>> 16) * 0xe995) << 16)) ^
-      /* Math.imul(h, m): */
-      ((h & 0xffff) * 0x5bd1e995 + (((h >>> 16) * 0xe995) << 16))
+  let i = 0
+  let len = length
+  for (; len >= 4; i++, len -= 4) {
+    let k = int32View[i]
+
+    k = Math.imul(k, m)
```

Comment:
`Math.imul` calls are better. It looks like a function call but the engine is free to transform this operation in a bitwise operation at the native level.

Using JS number operators (non-bitwise ones) is also risky. If a number overflows `2^31` in V8, it gets represented by a heap value (whereas below it's represented by a stack value, see https://v8.dev/blog/pointer-compression). Using `Math.imul` ensures the number stays as an `int32_t` value. This is also why I used an `Int32Array` to view the buffer, we absolutely want to keep numbers in the V8 Smi range.